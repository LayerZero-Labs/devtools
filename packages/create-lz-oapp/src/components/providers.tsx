import React, { useState } from "react"
import { QueryClient, QueryClientProvider } from "@tanstack/react-query"

export const Providers: React.FC<React.PropsWithChildren> = ({ children }) => {
    const [queryClient] = useState(() => new QueryClient())

    return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
}
