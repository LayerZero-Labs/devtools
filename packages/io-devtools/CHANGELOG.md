# @layerzerolabs/io-devtools

## 0.1.10

### Patch Changes

- 8b32a79: Use less strict version requirements for layerzero peer dependencies

## 0.1.9

### Patch Changes

- 938ac3d: Improve error reporting on UIntBigIntSchema

## 0.1.8

### Patch Changes

- 9af95ca: Add createWithAsyncLogger logging utility for async functions

## 0.1.7

### Patch Changes

- 8a4addf: Replace cli-table3 with table

## 0.1.6

### Patch Changes

- 2a36974: Add the ability to define configs as functions

## 0.1.5

### Patch Changes

- 54cf16e: Silence bigint-buffer warning

## 0.1.4

### Patch Changes

- 9e78abe: Update hardhat-deploy

## 0.1.3

### Patch Changes

- e766c86: Adding in Default Colors to print outs

## 0.1.2

### Patch Changes

- 253c79e: Rename description to hint for select prompts
- 5c58d69: Update @layerzerolabs dependencies to 2.1.3

## 0.1.1

### Patch Changes

- 40e8dba: Remove CommaSeparatedValuesSchema in favor of splitCommaSeparated; Move LogLevelSchema; Export isLogLevel

## 0.1.0

### Minor Changes

- 120adf1: Make packages public

## 0.0.7

### Patch Changes

- 81d7eee: Add promptForText utility

## 0.0.6

### Patch Changes

- 2b36ed9: Updating Table for Get OApp Config Task
- 2b36ed9: Rafactor some of the printers in io-devtools

## 0.0.5

### Patch Changes

- ed450c4: Add promptToSelectOne and promptToSelectMultiple utilities

## 0.0.4

### Patch Changes

- af8cc25: Update dependencies

## 0.0.3

### Patch Changes

- 5077bf1: Unmount the component when using printTable

## 0.0.2

### Patch Changes

- e0be8b7: Reexport Logger type
- 70646b4: Add missing dependencies

## 0.0.1

### Patch Changes

- b74afbe: Initial 0.0.1 version
