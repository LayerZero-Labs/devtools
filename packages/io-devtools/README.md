<p align="center">
  <a href="https://layerzero.network">
    <img alt="LayerZero" style="max-width: 500px" src="https://d3a2dpnnrypp5h.cloudfront.net/bridge-app/lz.png"/>
  </a>
</p>

<h1 align="center">@layerzerolabs/io-utils</h1>

<!-- The badges section -->
<p align="center">
  <!-- Shields.io NPM published package version -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM Version" src="https://img.shields.io/npm/v/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io NPM downloads -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="Downloads" src="https://img.shields.io/npm/dm/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io license badge -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM License" src="https://img.shields.io/npm/l/@layerzerolabs/io-utils"/></a>
</p>

## Installation

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
npm install @layerzerolabs/io-utils
```

## API Documentation

### Filesystem utilities

#### isDirectory(path)

Returns `true` if specified filesystem `path` points to a directory, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

#### isFile(path)

Returns `true` if specified filesystem `path` points to a file, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

#### isReadable(path)

Returns `true` if specified filesystem `path` can be read by the current user, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

### Standard input/output utilities

#### promptToContinue([message, defaultValue])

Asks the user whether they want to continue and reads the input from the CLI standard input. By default the question displayed is `Do you want to continue?` and the default response is `yes`

```typescript
const goahead = await promptToContinue();

// To ask a different question
const goahead = await promptToContinue("Are you sure?");

// To default the response to false, good for important and unsafe decisions
const goahead = await promptToContinue("Are you sure?", false);
```
