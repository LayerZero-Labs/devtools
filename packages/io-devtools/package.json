{
  "name": "@layerzerolabs/io-devtools",
  "version": "0.1.10",
  "description": "Utilities for working with I/O in LayerZero devtools",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LayerZero-Labs/devtools.git",
    "directory": "packages/io-devtools"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "types": "./dist/*/index.d.ts",
      "require": "./dist/*/index.js",
      "import": "./dist/*/index.mjs"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "swag"
  ],
  "scripts": {
    "prebuild": "tsc -noEmit",
    "build": "$npm_execpath tsup",
    "clean": "rm -rf dist",
    "dev": "$npm_execpath tsup --watch",
    "lint": "$npm_execpath eslint '**/*.{js,ts,json}'",
    "lint:fix": "eslint --fix '**/*.{js,ts,json}'",
    "test": "jest --ci"
  },
  "dependencies": {
    "chalk": "^4.1.2",
    "logform": "^2.6.0",
    "prompts": "^2.4.2",
    "table": "~6.8.2",
    "winston": "^3.11.0"
  },
  "devDependencies": {
    "@swc/core": "^1.4.0",
    "@swc/jest": "^0.2.36",
    "@types/ink-gradient": "^2.0.4",
    "@types/jest": "^29.5.12",
    "@types/prompts": "^2.4.9",
    "@types/react": "^17.0.75",
    "fast-check": "^3.15.1",
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "jest": "^29.7.0",
    "jest-extended": "^4.0.2",
    "react": "^17.0.2",
    "ts-node": "^10.9.2",
    "tslib": "~2.6.2",
    "tsup": "~8.0.1",
    "typescript": "^5.3.3",
    "zod": "^3.22.4"
  },
  "peerDependencies": {
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "react": "^17.0.2",
    "yoga-layout-prebuilt": "^1.9.6",
    "zod": "^3.22.4"
  },
  "peerDependenciesMeta": {
    "ink": {
      "optional": true
    },
    "ink-gradient": {
      "optional": true
    },
    "ink-table": {
      "optional": true
    },
    "react": {
      "optional": true
    },
    "yoga-layout-prebuilt": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
