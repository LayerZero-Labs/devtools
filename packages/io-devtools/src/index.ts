export * from './async'
export * from './config'
export * from './filesystem'
export * from './language'
export * from './stdio'
