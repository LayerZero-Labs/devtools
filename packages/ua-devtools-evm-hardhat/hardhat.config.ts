import 'hardhat-deploy'
import '../devtools-evm-hardhat/dist'
import { EndpointId } from '@layerzerolabs/lz-definitions'
import { HardhatUserConfig } from 'hardhat/types'

/**
 * This is a dummy hardhat config that enables us to test
 * hardhat functionality without mocking too much
 */
const config: HardhatUserConfig = {
    networks: {
        withEndpointId: {
            eid: EndpointId.ETHEREUM_MAINNET,
            url: 'no:///way',
            saveDeployments: false,
        },
        withoutEndpointId: {
            url: 'no:///way',
            saveDeployments: false,
        },
    },
}

export default config
