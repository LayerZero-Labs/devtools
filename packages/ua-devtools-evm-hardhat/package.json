{
  "name": "@layerzerolabs/ua-devtools-evm-hardhat",
  "version": "0.3.17",
  "description": "Hardhat utilities for working with LayerZero EVM contracts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LayerZero-Labs/devtools.git",
    "directory": "packages/ua-devtools-evm-hardhat"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./tasks": {
      "types": "./dist/tasks/index.d.ts",
      "require": "./dist/tasks/index.js",
      "import": "./dist/tasks/index.mjs"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "tasks",
    "dist"
  ],
  "scripts": {
    "prebuild": "tsc -noEmit",
    "build": "$npm_execpath tsup",
    "clean": "rm -rf dist",
    "dev": "$npm_execpath tsup --watch",
    "lint": "$npm_execpath eslint '**/*.{js,ts,json}'",
    "lint:fix": "eslint --fix '**/*.{js,ts,json}'"
  },
  "dependencies": {
    "p-memoize": "~4.0.4",
    "typescript": "^5.3.3"
  },
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/bytes": "^5.7.0",
    "@ethersproject/contracts": "^5.7.0",
    "@ethersproject/hash": "^5.7.0",
    "@layerzerolabs/devtools": "~0.3.15",
    "@layerzerolabs/devtools-evm": "~0.3.10",
    "@layerzerolabs/devtools-evm-hardhat": "~0.3.17",
    "@layerzerolabs/io-devtools": "~0.1.10",
    "@layerzerolabs/lz-definitions": "^2.3.3",
    "@layerzerolabs/lz-evm-messagelib-v2": "^2.3.3",
    "@layerzerolabs/lz-v2-utilities": "^2.3.3",
    "@layerzerolabs/protocol-devtools": "~0.3.8",
    "@layerzerolabs/protocol-devtools-evm": "~0.3.8",
    "@layerzerolabs/ua-devtools": "~0.3.13",
    "@layerzerolabs/ua-devtools-evm": "~0.3.12",
    "@swc/core": "^1.4.0",
    "@swc/jest": "^0.2.36",
    "@types/jest": "^29.5.12",
    "dotenv": "^16.4.1",
    "ethers": "^5.7.2",
    "fast-check": "^3.15.1",
    "hardhat": "^2.22.3",
    "hardhat-deploy": "^0.12.1",
    "jest": "^29.7.0",
    "ts-node": "^10.9.2",
    "tsup": "~8.0.1",
    "zod": "^3.22.4"
  },
  "peerDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/bytes": "^5.7.0",
    "@ethersproject/contracts": "^5.7.0",
    "@ethersproject/hash": "^5.7.0",
    "@layerzerolabs/devtools": "~0.3.15",
    "@layerzerolabs/devtools-evm": "~0.3.10",
    "@layerzerolabs/devtools-evm-hardhat": "~0.3.17",
    "@layerzerolabs/io-devtools": "~0.1.10",
    "@layerzerolabs/lz-definitions": "^2.3.3",
    "@layerzerolabs/protocol-devtools": "~0.3.8",
    "@layerzerolabs/protocol-devtools-evm": "~0.3.8",
    "@layerzerolabs/ua-devtools": "~0.3.13",
    "@layerzerolabs/ua-devtools-evm": "~0.3.12",
    "ethers": "^5.5.2",
    "hardhat": "^2.22.3",
    "hardhat-deploy": "^0.11.45"
  },
  "publishConfig": {
    "access": "public"
  }
}
