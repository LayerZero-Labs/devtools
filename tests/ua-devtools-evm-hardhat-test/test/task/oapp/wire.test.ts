import hre from 'hardhat'
import { isFile, promptToContinue } from '@layerzerolabs/io-devtools'
import { dirname, join, relative, resolve } from 'path'
import { TASK_LZ_OAPP_WIRE } from '@layerzerolabs/ua-devtools-evm-hardhat'
import { cwd } from 'process'
import { JsonRpcSigner } from '@ethersproject/providers'
import { deployContract, setupDefaultEndpointV2 } from '@layerzerolabs/test-setup-devtools-evm-hardhat'
import {
    createGnosisSignerFactory,
    createSignerFactory,
    SUBTASK_LZ_SIGN_AND_SEND,
} from '@layerzerolabs/devtools-evm-hardhat'
import { spawnSync } from 'child_process'

jest.mock('@layerzerolabs/io-devtools', () => {
    const original = jest.requireActual('@layerzerolabs/io-devtools')

    return {
        ...original,
        promptToContinue: jest.fn().mockRejectedValue('Not mocked'),
    }
})

jest.mock('@layerzerolabs/devtools-evm-hardhat', () => {
    const original = jest.requireActual('@layerzerolabs/devtools-evm-hardhat')

    return {
        ...original,
        createGnosisSignerFactory: jest.fn(original.createGnosisSignerFactory),
        createSignerFactory: jest.fn(original.createSignerFactory),
    }
})

const hreRunSpy = jest.spyOn(hre, 'run')
const promptToContinueMock = promptToContinue as jest.Mock
const createGnosisSignerFactoryMock = createGnosisSignerFactory as jest.Mock
const createSignerFactoryMock = createSignerFactory as jest.Mock

describe(`task ${TASK_LZ_OAPP_WIRE}`, () => {
    // Helper matcher object that checks for OmniPoint objects
    const expectOmniPoint = { address: expect.any(String), eid: expect.any(Number) }
    // Helper matcher object that checks for OmniTransaction objects
    const expectTransaction = { data: expect.any(String), point: expectOmniPoint, description: expect.any(String) }
    const expectTransactionWithReceipt = { receipt: expect.any(Object), transaction: expectTransaction }

    const CONFIGS_BASE_DIR = relative(cwd(), join(__dirname, '__data__', 'configs'))
    const configPathFixture = (fileName: string): string => {
        const path = join(CONFIGS_BASE_DIR, fileName)

        expect(isFile(path)).toBeTruthy()

        return path
    }

    beforeAll(async () => {
        await deployContract('EndpointV2')
        await setupDefaultEndpointV2()
    })

    beforeEach(async () => {
        promptToContinueMock.mockReset()
        createGnosisSignerFactoryMock.mockClear()
        createSignerFactoryMock.mockClear()
        hreRunSpy.mockClear()
    })

    describe('with invalid configs', () => {
        beforeAll(async () => {
            await deployContract('OApp')
        })

        it('should fail if the config file does not exist', async () => {
            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig: './does-not-exist.js' })).rejects.toMatchSnapshot()
        })

        it('should fail if the config file is not a file', async () => {
            const oappConfig = dirname(configPathFixture('invalid.config.empty.json'))

            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail if the config file is not a valid JSON or JS file', async () => {
            const oappConfig = 'README.md'

            expect(isFile(oappConfig)).toBeTruthy()

            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with an empty JSON file', async () => {
            const oappConfig = configPathFixture('invalid.config.empty.json')

            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with an empty JS file', async () => {
            const oappConfig = configPathFixture('invalid.config.empty.js')

            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with a malformed JS file (001)', async () => {
            const oappConfig = configPathFixture('invalid.config.001.js')

            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })).rejects.toMatchSnapshot()
        })

        it('should fail with a misconfigured file (001)', async () => {
            const oappConfig = configPathFixture('valid.config.misconfigured.001.js')

            await expect(hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })).rejects.toMatchSnapshot()
        })
    })

    describe('with valid configs', () => {
        beforeEach(async () => {
            await deployContract('OApp')
        })

        it('should accept a path without an extension', async () => {
            const oappConfig = configPathFixture('valid.config.empty.js')

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should exit if there is nothing to wire', async () => {
            const oappConfig = configPathFixture('valid.config.empty.js')

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should work with absolute paths', async () => {
            const oappConfigAbsolute = configPathFixture('valid.config.empty.js')
            const oappConfig = resolve(oappConfigAbsolute)

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should work with typescript', async () => {
            const oappConfig = configPathFixture('valid.config.empty.ts')

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should have debug output if requested (so called eye test, check the test output)', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(false)

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, logLevel: 'debug' })

            expect(promptToContinueMock).toHaveBeenCalledTimes(2)
        })

        it('should not ask the user for input if --ci flag is truthy and execute all transactions', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(false)

            const result = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, ci: true })

            expect(result).toEqual([[expectTransactionWithReceipt, expectTransactionWithReceipt], [], []])
            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should ask the user whether they want to see the transactions & continue', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(true)

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            expect(promptToContinueMock).toHaveBeenCalledTimes(2)
        })

        it('should return a list of pending transactions if the user decides not to continue', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock.mockResolvedValue(false)

            const [successful, errors, pending] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            expect(successful).toEqual([])
            expect(errors).toEqual([])
            expect(pending).toHaveLength(2)
            expect(promptToContinueMock).toHaveBeenCalledTimes(2)
        })

        it('should return a list of pending transactions if running in dry run mode', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            const [successful, errors, pending] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, dryRun: true })

            expect(successful).toEqual([])
            expect(errors).toEqual([])
            expect(pending).toHaveLength(2)
            expect(promptToContinueMock).not.toHaveBeenCalled()
        })

        it('should return a list of transactions if the user decides to continue', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            const [successful, errors] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

            const expectTransactionWithReceipt = { receipt: expect.any(Object), transaction: expect.any(Object) }

            expect(successful).toEqual([expectTransactionWithReceipt, expectTransactionWithReceipt])
            expect(errors).toEqual([])
        })

        it('should use a named signer if a signer name is passed', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')
            const signer = { type: 'named', name: 'wombat' }

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, signer })

            expect(createSignerFactoryMock).toHaveBeenCalledOnce()
            expect(createSignerFactoryMock).toHaveBeenCalledWith(signer)
        })

        it('should use a named safe signer if a signer name is passed', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')
            const signer = { type: 'named', name: 'wombat' }

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, safe: true, signer })

            expect(createGnosisSignerFactory).toHaveBeenCalledOnce()
            expect(createGnosisSignerFactory).toHaveBeenCalledWith(signer)
        })

        it('should use gnosis safe signer if --safe flag is passed', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, safe: true })

            expect(createGnosisSignerFactory).toHaveBeenCalledOnce()

            // Get the created gnosis signer factory
            const createSigner = createGnosisSignerFactoryMock.mock.results[0]?.value
            expect(typeof createSigner).toBe('function')

            // Now we check that the sign and send subtask has been called with the correct signer factory
            expect(hreRunSpy).toHaveBeenCalledWith(
                SUBTASK_LZ_SIGN_AND_SEND,
                {
                    transactions: expect.any(Array),
                    ci: false,
                    createSigner,
                },
                {},
                undefined
            )
        })

        it('should use gnosis safe signer with index if --safe flag and --signer are passed', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')
            const [signer] = await hre.getUnnamedAccounts()

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, safe: true, signer: { type: 'address', address: signer } })

            expect(createGnosisSignerFactory).toHaveBeenCalledOnce()
            expect(createGnosisSignerFactory).toHaveBeenCalledWith({ type: 'address', address: signer })

            // Get the created gnosis signer factory
            const createSigner = createGnosisSignerFactoryMock.mock.results[0]?.value
            expect(typeof createSigner).toBe('function')

            // Now we check that the sign and send subtask has been called with the correct signer factory
            expect(hreRunSpy).toHaveBeenCalledWith(
                SUBTASK_LZ_SIGN_AND_SEND,
                {
                    transactions: expect.any(Array),
                    ci: false,
                    createSigner,
                },
                {},
                undefined
            )
        })

        it('should use signer index if --signer is passed', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, signer: 0 })

            expect(createSignerFactory).toHaveBeenCalledOnce()
            expect(createSignerFactory).toHaveBeenCalledWith(0)

            // Get the created signer factory
            const createSigner = createSignerFactoryMock.mock.results[0]?.value
            expect(typeof createSigner).toBe('function')

            // Now we check that the sign and send subtask has been called with the correct signer factory
            expect(hreRunSpy).toHaveBeenCalledWith(
                SUBTASK_LZ_SIGN_AND_SEND,
                {
                    transactions: expect.any(Array),
                    ci: false,
                    createSigner,
                },
                {},
                undefined
            )
        })

        it('should use signer address if --signer is passed', async () => {
            const oappConfig = configPathFixture('valid.config.connected.js')
            const [signer] = await hre.getUnnamedAccounts()

            promptToContinueMock
                .mockResolvedValueOnce(false) // We don't want to see the list
                .mockResolvedValueOnce(true) // We want to continue

            await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, signer })

            expect(createSignerFactory).toHaveBeenCalledOnce()
            expect(createSignerFactory).toHaveBeenCalledWith(signer)

            // Get the created signer factory
            const createSigner = createSignerFactoryMock.mock.results[0]?.value
            expect(typeof createSigner).toBe('function')

            // Now we check that the sign and send subtask has been called with the correct signer factory
            expect(hreRunSpy).toHaveBeenCalledWith(
                SUBTASK_LZ_SIGN_AND_SEND,
                {
                    transactions: expect.any(Array),
                    ci: false,
                    createSigner,
                },
                {},
                undefined
            )
        })

        describe('if a transaction fails', () => {
            let sendTransactionMock: jest.SpyInstance

            beforeEach(() => {
                sendTransactionMock = jest.spyOn(JsonRpcSigner.prototype, 'sendTransaction')
            })

            afterEach(() => {
                sendTransactionMock.mockRestore()
            })

            it('should set the exit code to 1', async () => {
                const error = new Error('Oh god dammit')

                // We want to make the transaction fail
                sendTransactionMock.mockRejectedValue(error)

                const oappConfig = configPathFixture('valid.config.connected.js')
                await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, ci: true })

                expect(process.exitCode).toBe(1)
            })

            it('should return a list of failed transactions in the CI mode', async () => {
                const error = new Error('Oh god dammit')

                // We want to make the transaction fail
                sendTransactionMock.mockRejectedValue(error)

                const oappConfig = configPathFixture('valid.config.connected.js')
                const [successful, errors, pending] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, ci: true })

                // The transactions are being grouped by chain and signed in parallel
                // so we expect one failure per chain
                expect(errors).toEqual([
                    {
                        error,
                        transaction: expectTransaction,
                    },
                    {
                        error,
                        transaction: expectTransaction,
                    },
                ])

                // Since we failed on the first transaction, we expect
                // all the transaction to still be pending and none of them to be successful
                expect(successful).toEqual([])
                expect(pending).toEqual([expectTransaction, expectTransaction])
            })

            it('should ask the user to retry if not in the CI mode', async () => {
                const error = new Error('Oh god dammit')

                // Mock the first sendTransaction call to reject, the rest should use the original implementation
                //
                // This way we simulate a situation in which the first call would fail but then the user retries, it would succeed
                sendTransactionMock.mockRejectedValueOnce(error)

                // In the non-CI mode we need to answer the prompts
                promptToContinueMock
                    .mockResolvedValueOnce(false) // We don't want to see the list of transactions
                    .mockResolvedValueOnce(true) // We want to continue
                    .mockResolvedValueOnce(true) // We want to see the list of failed transactions
                    .mockResolvedValueOnce(true) // We want to retry

                const oappConfig = configPathFixture('valid.config.connected.js')
                const [successful, errors, pending] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

                // Check that the user has been asked to retry
                expect(promptToContinueMock).toHaveBeenCalledWith(`Would you like to preview the failed transactions?`)
                expect(promptToContinueMock).toHaveBeenCalledWith(`Would you like to retry?`, true)

                // After retrying, the signer should not fail anymore
                expect(successful).toEqual([expectTransactionWithReceipt, expectTransactionWithReceipt])
                expect(errors).toEqual([])
                expect(pending).toEqual([])
            })

            it('should not retry if the user decides not to if not in the CI mode', async () => {
                const error = new Error('Oh god dammit')

                // Mock the first sendTransaction call to reject, the rest should use the original implementation
                //
                // This way we simulate a situation in which the first call would fail but then the user retries, it would succeed
                sendTransactionMock.mockRejectedValueOnce(error)

                // In the non-CI mode we need to answer the prompts
                promptToContinueMock
                    .mockResolvedValueOnce(false) // We don't want to see the list of transactions
                    .mockResolvedValueOnce(true) // We want to continue
                    .mockResolvedValueOnce(false) // We don't want to see the list of failed transactions
                    .mockResolvedValueOnce(false) // We don't want to retry

                const oappConfig = configPathFixture('valid.config.connected.js')
                const [successful, errors, pending] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

                // Check that the user has been asked to retry
                expect(promptToContinueMock).toHaveBeenCalledWith(`Would you like to preview the failed transactions?`)
                expect(promptToContinueMock).toHaveBeenCalledWith(`Would you like to retry?`, true)

                // Check that we got the failures back
                expect(errors).toEqual([
                    {
                        error,
                        transaction: expectTransaction,
                    },
                ])

                // Since we failed on the first transaction (on one chain only),
                // we expect one transaction on the other chain to go though just fine
                // and the failed one to appear in the pending array
                expect(successful).toEqual([expectTransactionWithReceipt])
                expect(pending).toEqual([expectTransaction])
            })

            it('should not retry successful transactions', async () => {
                const error = new Error('Oh god dammit')

                // Mock the second & third sendTransaction call to reject
                //
                // This way we simulate a situation in which the first call goes through,
                // then the second and third calls reject
                sendTransactionMock
                    .mockImplementationOnce(sendTransactionMock.getMockImplementation()!)
                    .mockRejectedValueOnce(error)
                    .mockRejectedValueOnce(error)

                // In the non-CI mode we need to answer the prompts
                promptToContinueMock
                    .mockResolvedValueOnce(false) // We don't want to see the list of transactions
                    .mockResolvedValueOnce(true) // We want to continue
                    .mockResolvedValueOnce(true) // We want to see the list of failed transactions
                    .mockResolvedValueOnce(true) // We want to retry
                    .mockResolvedValueOnce(true) // We want to see the list of failed transactions
                    .mockResolvedValueOnce(true) // We want to retry

                const oappConfig = configPathFixture('valid.config.connected.js')
                const [successful, errors, pending] = await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig })

                // Check that the user has been asked to retry
                expect(promptToContinueMock).toHaveBeenCalledWith(`Would you like to preview the failed transactions?`)
                expect(promptToContinueMock).toHaveBeenCalledWith(`Would you like to retry?`, true)

                // After retrying, the signer should not fail anymore
                expect(successful).toEqual([expectTransactionWithReceipt, expectTransactionWithReceipt])
                expect(errors).toEqual([])
                expect(pending).toEqual([])

                expect(sendTransactionMock).toHaveBeenCalledTimes(
                    // The first successful call
                    1 +
                        // The first failed call
                        1 +
                        // The retry of the failed call
                        1 +
                        // The retry of the failed call
                        1
                )
            })
        })
    })

    describe('expectations', () => {
        const EXPECTATIONS_DIRECTORY = join('test', 'task', 'oapp', 'config.test.expectations')
        const expectationPath = (name: string) => join(EXPECTATIONS_DIRECTORY, `${name}.exp`)
        const runExpect = (name: string) =>
            spawnSync(expectationPath(name), {
                encoding: 'utf8',
                stdio: 'inherit',
            })

        beforeEach(async () => {
            await deployContract('EndpointV2', true)
            await setupDefaultEndpointV2()
        })

        it('should use custom config loading task', async () => {
            const result = runExpect('config-custom-config-loading')

            expect(result.status).toBe(0)
        })
    })
})
