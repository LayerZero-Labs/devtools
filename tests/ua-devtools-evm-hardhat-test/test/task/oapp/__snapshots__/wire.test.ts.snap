// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task/oapp/wire with invalid configs should fail if the config file does not exist 1`] = `[Error: Unable to read config file '/app/tests/ua-devtools-evm-hardhat-test/does-not-exist.js'. Check that the file exists and is readable to your terminal user]`;

exports[`task/oapp/wire with invalid configs should fail if the config file is not a file 1`] = `[Error: Unable to read config file '/app/tests/ua-devtools-evm-hardhat-test/test/task/oapp'. Check that the file exists and is readable to your terminal user]`;

exports[`task/oapp/wire with invalid configs should fail if the config file is not a valid JSON or JS file 1`] = `[Error: Unable to read config file '/app/tests/ua-devtools-evm-hardhat-test/README.md': SyntaxError: Unexpected token '<']`;

exports[`task/oapp/wire with invalid configs should fail with a malformed JS file (001) 1`] = `
[Error: Config from file '/app/tests/ua-devtools-evm-hardhat-test/test/task/oapp/__data__/configs/invalid.config.001.js' is malformed. Please fix the following errors:

Property 'contracts.0.contract': Invalid input
Property 'contracts.1.contract': Invalid input
Property 'connections': Required]
`;

exports[`task/oapp/wire with invalid configs should fail with a misconfigured file (001) 1`] = `[Error: Config from file '/app/tests/ua-devtools-evm-hardhat-test/test/task/oapp/__data__/configs/valid.config.misconfigured.001.js' is invalid: AssertionError [ERR_ASSERTION]: Could not find a deployment for contract 'NonExistent']`;

exports[`task/oapp/wire with invalid configs should fail with an empty JS file 1`] = `
[Error: Config from file '/app/tests/ua-devtools-evm-hardhat-test/test/task/oapp/__data__/configs/invalid.config.empty.js' is malformed. Please fix the following errors:

Property 'contracts': Required
Property 'connections': Required]
`;

exports[`task/oapp/wire with invalid configs should fail with an empty JSON file 1`] = `[Error: Unable to read config file '/app/tests/ua-devtools-evm-hardhat-test/test/task/oapp/__data__/configs/invalid.config.empty.json': SyntaxError: Unexpected end of JSON input]`;
